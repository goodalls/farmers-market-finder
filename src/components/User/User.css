@import url("https://fonts.googleapis.com/css?family=Open+Sans:400,700,800");
.user {
  border: 2px solid #9BC47e;
  border-radius: 10px;
  display: flex;
  flex-direction: column; }

.user input {
  font-size: 1em;
  height: 25px;
  margin: 5px; }

.user h4 {
  padding: 0px;
  margin: 0px; }

#name {
  align-self: flex-start; }

#password {
  align-self: flex-start; }

#submit {
  align-self: flex-start; }

button.user {
  font-size: 1em; }
