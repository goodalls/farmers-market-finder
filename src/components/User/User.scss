@import '../../styles/colors.scss';

.user {
  border: 2px solid $primary-green;
  border-radius: 10px;
  display: flex;
  flex-direction: column;
}

.user input {
  // background-color: transparent;
  // border-radius: 10px;
  // border: 2px solid #9BC47e;
  // color: #D97426;
  font-size: 1em;
  height: 25px;
  // margin: 20px 10px;
  // width: 50%;
  // text-align: center;
  // outline: none;
  margin: 5px;
}

.user h4 {
  padding: 0px;
  margin: 0px; !important
}

#name {
  align-self: flex-start;
}

#password {
  align-self: flex-start;
}

#submit {
  align-self: flex-start;
}

.user input {
}

button.user {
  background-color: transparent;
  border-radius: 10px;
  border: 2px solid #9BC47e;
  color: #D97426;
  font-size: 1em;
  margin: 20px 10px;
  width: 50%;
  outline: none;
}
