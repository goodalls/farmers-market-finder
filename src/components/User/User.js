import React, { Component } from 'react';
import * as actions from '../../actions/actions';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import './User.css';

export class User extends Component {
  constructor(props) {
    super(props);
    this.state = {
      status: '',
      name: '',
      email: '',
      password: '',
      favorites: []
    };
  }
  componentDidMount() {
    this.userUpdater();
  }

  userUpdater() {
    if (localStorage.user) {
      const get = localStorage.getItem('user');
      const user = JSON.parse(get);
      this.setState({ ...user, status: 'LOGGED_IN' });
      this.props.loginUser({ ...user, status: 'LOGGED_IN' });
    }
  }

  handleInputs = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  loginUser() {
    return (
      <div className="user">
        <h4>Login</h4>
        <form onSubmit={this.saveUser}>
          <label htmlFor="name" />
          <input
            id="name"
            name="name"
            type="text"
            placeholder="Name"
            onChange={this.handleInputs}
            value={this.state.name}
          />
          <label htmlFor="password" />
          <input
            id="password"
            name="password"
            type="password"
            placeholder="Password"
            onChange={this.handleInputs}
            value={this.state.password}
          />
          <label htmlFor="submit" />
          <input id="submit" type="submit" value="Submit" />
          <label htmlFor="create-acct" />
          <input
            id="create-acct"
            type="submit"
            value="Create Account"
            onClick={this.createUser}
          />
        </form>
      </div>
    );
  }

  logoutUser() {
    return (
      <div className="user">
        <h4>Welcome back {this.state.name}</h4>
        <Link to="/favorite">
          <button className="favorites">
            favorites {this.props.user.favorites.length}
          </button>
        </Link>
        <div>
          <button
            onClick={() => {
              localStorage.removeItem('user');
              this.setState({
                status: '',
                name: '',
                email: '',
                password: '',
                favorites: []
              });
              this.props.logOutUser();
              // this.props.history.push('/');
            }}
          >
            Log Out
          </button>
        </div>
      </div>
    );
  }

  createUser = () => {
    if (this.state.status !== 'CREATE_USER') {
      this.setState({ status: 'CREATE_USER' });
      this.props.updateUser({ status: 'CREATE_USER' });
    } else {
      return (
        <div className="user">
          <form onSubmit={this.saveUser}>
            <label htmlFor="name" />
            <input
              id="name"
              name="name"
              type="text"
              placeholder="Name"
              onChange={this.handleInputs}
              value={this.state.name}
            />
            <label htmlFor="email" />
            <input
              id="email"
              name="email"
              type="text"
              placeholder="Email"
              onChange={this.handleInputs}
              value={this.state.email}
            />
            <label htmlFor="password" />
            <input
              id="password"
              name="password"
              type="text"
              placeholder="Password"
              onChange={this.handleInputs}
              value={this.state.password}
            />
            <label htmlFor="submit" />
            <input id="submit" type="submit" value="Submit" />
          </form>
        </div>
      );
    }
  };

  saveUser = () => {
    const stringify = JSON.stringify(this.state);
    localStorage.setItem('user', stringify);
    this.setState({ status: 'LOGGED_IN' });
    this.props.loginUser(this.state);
  };

  renderCheck = user => {
    switch (user.status) {
    case 'LOGGED_IN':
      return this.logoutUser();
    case 'LOGGED_OUT':
      return this.loginUser();
    case 'CREATE_USER':
      return this.createUser();
    default:
      return this.loginUser();
    }
  };
  render() {
    return <div>{this.renderCheck(this.state)}</div>;
  }
}

User.propTypes = {
  history: PropTypes.object,
  logOutUser: PropTypes.func,
  loginUser: PropTypes.func,
  updateUser: PropTypes.func,
  user: PropTypes.object
};

export const mapStateToProps = state => ({
  user: state.user
});

export const mapDispatchToProps = dispatch => ({
  loginUser: user => dispatch(actions.loginUser(user)),
  logOutUser: () => dispatch(actions.logOutUser()),
  updateUser: user => dispatch(actions.updateUser(user))
});

export default connect(mapStateToProps, mapDispatchToProps)(User);
