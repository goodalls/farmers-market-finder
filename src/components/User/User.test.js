import React from 'react';
import {shallow} from 'enzyme';
import {User, mapStateToProps, mapDispatchToProps} from './User';


describe('USER', () => {
  let wrapper;
  beforeAll(()=> {
    const localStorageMock = {
      getItem: jest.fn(),
      setItem: jest.fn(),
      clear: jest.fn()
    };
    global.localStorage = localStorageMock;
  });

  beforeEach(()=>{
    wrapper = shallow(<User history={({})} logOutUser={jest.fn()} loginUser={jest.fn()} updateUser={jest.fn()} user={({})} />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('userUpdater', () => {
    it('should set state of user from localstorage', () => {
      const localStorageMock = {
        getItem: jest.fn().mockReturnValue({name: 'tom'}),
        setItem: jest.fn(),
        clear: jest.fn()
      };
      const expected = {
        "email": "", 
        "favorites":[], 
        "name": "", 
        "password": "", 
        "status": ""
      };
      global.localStorage = localStorageMock;
      wrapper.instance().userUpdater();
      wrapper.update();
      expect(wrapper.state()).toEqual(expected);
    });

    it.skip('should call localStorage.getItem', () => {
      const localStorageMock = {
        getItem: jest.fn(),
        setItem: jest.fn(),
        clear: jest.fn()
      };
      global.localStorage = localStorageMock;
      wrapper.instance().userUpdater();
      expect(localStorage.getItem).toHaveBeenCalled();
    });
  });
  
  describe('handleInputs', () => {
    it('should set state from the name input', () => {
      const mockEvent = {target: {name: 'name', value: 'the man'}};
      wrapper.instance().handleInputs(mockEvent);
      expect(wrapper.state('name')).toEqual('the man');
    });

    it('should set state from the password input', () => {
      const mockEvent = {target: {name: 'password', value: 'password'}};
      wrapper.instance().handleInputs(mockEvent);
      expect(wrapper.state('password')).toEqual('password');
    });

    it('should set state from the email input', () => {
      const mockEvent = {target: {name: 'email', value: 'me@me.com'}};
      wrapper.instance().handleInputs(mockEvent);
      expect(wrapper.state('email')).toEqual('me@me.com');
    });
  });

  describe('saveUser', () => {
    it('should call localStorage setItem', () => {
      wrapper.instance().saveUser();
      expect(localStorage.setItem()).toHaveBeenCalled();
    });
  });
  
  

  describe('MSTP and MDTP', () => {
    it('should map user info to props', () => {
      const mockState = {
        user: {}
      };
      const expected = {};
      const mapped = mapStateToProps(mockState);
      expect(mapped.user).toEqual(expected);
    });

    it('should call dispatch on loginUser', () => {
      const mockDispatch = jest.fn();
      const mapped = mapDispatchToProps(mockDispatch);
      mapped.loginUser();
      expect(mockDispatch).toHaveBeenCalled();
    });

    it('should call dispatch on loginUser', () => {
      const mockDispatch = jest.fn();
      const mapped = mapDispatchToProps(mockDispatch);
      mapped.logOutUser();
      expect(mockDispatch).toHaveBeenCalled();
    });

    it('should call dispatch on loginUser', () => {
      const mockDispatch = jest.fn();
      const mapped = mapDispatchToProps(mockDispatch);
      mapped.updateUser();
      expect(mockDispatch).toHaveBeenCalled();
    });
  });
  

});
