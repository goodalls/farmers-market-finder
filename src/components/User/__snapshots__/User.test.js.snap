// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`USER should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <User
    history={Object {}}
    logOutUser={[Function]}
    loginUser={[Function]}
    updateUser={[Function]}
    user={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="user"
>
        <h4>
                Login
        </h4>
        <form
                onSubmit={[Function]}
        >
                <label
                        htmlFor="name"
                />
                <input
                        id="name"
                        name="name"
                        onChange={[Function]}
                        placeholder="Name"
                        type="text"
                        value=""
                />
                <label
                        htmlFor="password"
                />
                <input
                        id="password"
                        name="password"
                        onChange={[Function]}
                        placeholder="Password"
                        type="password"
                        value=""
                />
                <label
                        htmlFor="submit"
                />
                <input
                        id="submit"
                        type="submit"
                        value="Submit"
                />
                <label
                        htmlFor="create-acct"
                />
                <input
                        id="create-acct"
                        onClick={[Function]}
                        type="submit"
                        value="Create Account"
                />
        </form>
</div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h4>
            Login
</h4>,
          <form
            onSubmit={[Function]}
>
            <label
                        htmlFor="name"
            />
            <input
                        id="name"
                        name="name"
                        onChange={[Function]}
                        placeholder="Name"
                        type="text"
                        value=""
            />
            <label
                        htmlFor="password"
            />
            <input
                        id="password"
                        name="password"
                        onChange={[Function]}
                        placeholder="Password"
                        type="password"
                        value=""
            />
            <label
                        htmlFor="submit"
            />
            <input
                        id="submit"
                        type="submit"
                        value="Submit"
            />
            <label
                        htmlFor="create-acct"
            />
            <input
                        id="create-acct"
                        onClick={[Function]}
                        type="submit"
                        value="Create Account"
            />
</form>,
        ],
        "className": "user",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Login",
          },
          "ref": null,
          "rendered": "Login",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label
                htmlFor="name"
/>,
              <input
                id="name"
                name="name"
                onChange={[Function]}
                placeholder="Name"
                type="text"
                value=""
/>,
              <label
                htmlFor="password"
/>,
              <input
                id="password"
                name="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
/>,
              <label
                htmlFor="submit"
/>,
              <input
                id="submit"
                type="submit"
                value="Submit"
/>,
              <label
                htmlFor="create-acct"
/>,
              <input
                id="create-acct"
                onClick={[Function]}
                type="submit"
                value="Create Account"
/>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "htmlFor": "name",
              },
              "ref": null,
              "rendered": null,
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "name",
                "name": "name",
                "onChange": [Function],
                "placeholder": "Name",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "htmlFor": "password",
              },
              "ref": null,
              "rendered": null,
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "password",
                "name": "password",
                "onChange": [Function],
                "placeholder": "Password",
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "htmlFor": "submit",
              },
              "ref": null,
              "rendered": null,
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "submit",
                "type": "submit",
                "value": "Submit",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "htmlFor": "create-acct",
              },
              "ref": null,
              "rendered": null,
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "create-acct",
                "onClick": [Function],
                "type": "submit",
                "value": "Create Account",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="user"
>
          <h4>
                    Login
          </h4>
          <form
                    onSubmit={[Function]}
          >
                    <label
                              htmlFor="name"
                    />
                    <input
                              id="name"
                              name="name"
                              onChange={[Function]}
                              placeholder="Name"
                              type="text"
                              value=""
                    />
                    <label
                              htmlFor="password"
                    />
                    <input
                              id="password"
                              name="password"
                              onChange={[Function]}
                              placeholder="Password"
                              type="password"
                              value=""
                    />
                    <label
                              htmlFor="submit"
                    />
                    <input
                              id="submit"
                              type="submit"
                              value="Submit"
                    />
                    <label
                              htmlFor="create-acct"
                    />
                    <input
                              id="create-acct"
                              onClick={[Function]}
                              type="submit"
                              value="Create Account"
                    />
          </form>
</div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h4>
              Login
</h4>,
            <form
              onSubmit={[Function]}
>
              <label
                            htmlFor="name"
              />
              <input
                            id="name"
                            name="name"
                            onChange={[Function]}
                            placeholder="Name"
                            type="text"
                            value=""
              />
              <label
                            htmlFor="password"
              />
              <input
                            id="password"
                            name="password"
                            onChange={[Function]}
                            placeholder="Password"
                            type="password"
                            value=""
              />
              <label
                            htmlFor="submit"
              />
              <input
                            id="submit"
                            type="submit"
                            value="Submit"
              />
              <label
                            htmlFor="create-acct"
              />
              <input
                            id="create-acct"
                            onClick={[Function]}
                            type="submit"
                            value="Create Account"
              />
</form>,
          ],
          "className": "user",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Login",
            },
            "ref": null,
            "rendered": "Login",
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <label
                  htmlFor="name"
/>,
                <input
                  id="name"
                  name="name"
                  onChange={[Function]}
                  placeholder="Name"
                  type="text"
                  value=""
/>,
                <label
                  htmlFor="password"
/>,
                <input
                  id="password"
                  name="password"
                  onChange={[Function]}
                  placeholder="Password"
                  type="password"
                  value=""
/>,
                <label
                  htmlFor="submit"
/>,
                <input
                  id="submit"
                  type="submit"
                  value="Submit"
/>,
                <label
                  htmlFor="create-acct"
/>,
                <input
                  id="create-acct"
                  onClick={[Function]}
                  type="submit"
                  value="Create Account"
/>,
              ],
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "htmlFor": "name",
                },
                "ref": null,
                "rendered": null,
                "type": "label",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "id": "name",
                  "name": "name",
                  "onChange": [Function],
                  "placeholder": "Name",
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "htmlFor": "password",
                },
                "ref": null,
                "rendered": null,
                "type": "label",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "id": "password",
                  "name": "password",
                  "onChange": [Function],
                  "placeholder": "Password",
                  "type": "password",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "htmlFor": "submit",
                },
                "ref": null,
                "rendered": null,
                "type": "label",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "id": "submit",
                  "type": "submit",
                  "value": "Submit",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "htmlFor": "create-acct",
                },
                "ref": null,
                "rendered": null,
                "type": "label",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "id": "create-acct",
                  "onClick": [Function],
                  "type": "submit",
                  "value": "Create Account",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "form",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
